File: c++/121.cpp
// 121. Best Time to Buy and Sell Stock
// Say you have an array for which the ith element is the price of a given stock on day i.
//
// If you were only permitted to complete at most one transaction (i.e., buy one and sell one share of the stock), design an algorithm to find the maximum profit.
//
// Note that you cannot sell a stock before you buy one.
//
// Example 1:
//   Input: [7,1,5,3,6,4]
//   Output: 5
//   Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
//                Not 7-1 = 6, as selling price needs to be larger than buying price.
// Example 2:
//   Input: [7,6,4,3,1]
//   Output: 0
// Explanation: In this case, no transaction is done, i.e. max profit = 0.

// 368 ms, faster than 7.56%
class Solution {
public:
  int maxProfit(vector<int>& prices) {
    int len = prices.size();
    if (len < 2) return 0;
    int maxProfit = prices[1] - prices[0];
    for (int i = 0; i < len; i++) {
      for (int j = i+1; j < len; j++) {
        int temp = prices[j] - prices[i];
        if (temp < 0) {
          continue;
        } else if (temp > maxProfit) {
          maxProfit = temp;
        }
      }
    }
    if(maxProfit < 0) return 0;
    return maxProfit;
  }
};

// 4 ms, faster than 99.49%
// This method only works when the revenue is set to be greater than 0. That is when there is negative revenue, no transaction is done.
class Solution {
public:
  int maxProfit(vector<int>& prices) {
    int len = prices.size();
    if (len < 2) return 0;
    int min = prices[0];
    int maxProfit = 0;
    for (int i = 0; i < len; i++) {
      if (prices[i] < min) {
        min = prices[i];
      } else {
        int temp = prices[i] - min;
        if (temp > maxProfit) {
          maxProfit = temp;
        }
      }
    }
    return maxProfit;
  }
};

File: go/121.go
// 121. Best Time to Buy and Sell Stock

// You are given an array prices where prices[i] is the price of a given stock on the ith day.

// You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

// Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

 

// Example 1:

// Input: prices = [7,1,5,3,6,4]
// Output: 5
// Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
// Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
// Example 2:

// Input: prices = [7,6,4,3,1]
// Output: 0
// Explanation: In this case, no transactions are done and the max profit = 0.
 

// Constraints:

// 1 <= prices.length <= 105
// 0 <= prices[i] <= 104

// Runtime 0 ms Beats 100.00%
// Memory 10.52 MB Beats 31.83%
func maxProfit(prices []int) int {
    localMin := prices[0]
    localMax := prices[0]
    profit := 0
    for i := 1; i < len(prices); i++ {
        if localMax < prices[i] {
            localMax = prices[i]
            profit = max(profit, localMax - localMin)
        } else if localMin > prices[i] {
            localMin = prices[i]
            localMax = prices[i]
        }
    }
    return profit
}